sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala

```diff
@@ -344,4 +346,36 @@ private[csv] object UnivocityParser {
       CSVUtils.filterCommentAndEmpty(linesWithoutHeader, options)
     filteredLines.flatMap(line => parser.parse(line))
   }
+
+  /**
+   * Parses a `Dataset` that contains CSV strings and turns it into an `RDD` of rows.
+   */
+  def tokenizeDataset(
+      csvDataset: Dataset[String],
+      maybeFirstLine: Option[String],
+      options: CSVOptions): RDD[Array[String]] = {
+    val filtered = CSVUtils.filterCommentAndEmpty(csvDataset, options)
+    val linesWithoutHeader = maybeFirstLine.map { firstLine =>
+      filtered.rdd.mapPartitions(CSVUtils.filterHeaderLine(_, firstLine, options))
+    }.getOrElse(filtered.rdd)
+
+    linesWithoutHeader.mapPartitions { iter =>
+      val parser = new CsvParser(options.asParserSettings)
+      iter.map(line => parser.parseLine(line))
+    }
+  }
+
+  /**
+   * Parses a `Dataset` that contains CSV strings and turns it into an `RDD` of rows.
+   */
+  def parseDataset(
+      csvDataset: Dataset[String],
+      schema: StructType,
+      maybeFirstLine: Option[String],
+      options: CSVOptions): RDD[InternalRow] = {
+    tokenizeDataset(csvDataset, maybeFirstLine, options).mapPartitions { iter =>
+      val parser = new UnivocityParser(schema, options)
+      iter.flatMap(line => parser.convert(line))
+    }
+  }
```

Comment:
If you are not sure or it looks painful to review, let me take out all the changes and put those into `DataFrameReader.csv` for now.